include/lo2s/perf/event.hpp

```diff
@@ -0,0 +1,291 @@
+
+    /**
+     * returns an opened instance of any Event object
+     */
+    EventGuard open(std::variant<Cpu, Thread> location, int cgroup_fd = -1);
+    EventGuard open(ExecutionScope location, int cgroup_fd = -1);
+
+    /**
+     * returns an opened instance of a Event object after formating it as a leader Event
+     */
+    EventGuard open_as_group_leader(ExecutionScope location, int cgroup_fd = -1);
+
+    const Availability& get_availability() const
+    {
+        return availability_;
+    };
+
+    std::string get_name() const
+    {
+        return name_;
+    }
+
+    std::set<Cpu> get_cpus() const
+    {
+        return cpus_;
+    }
+
+    std::string get_unit() const
+    {
+        return unit_;
+    }
+
+    perf_event_attr& get_attr()
+    {
+        return attr_;
+    }
+
+    double get_scale() const
+    {
+        return scale_;
+    }
+
+    void set_scale(double scale)
+    {
+        scale_ = scale;
+    }
+
+    void set_unit(std::string unit)
+    {
+        unit_ = unit;
+    }
+
+    void set_clock_attrs([[maybe_unused]] const bool& use_clockid,
+                         [[maybe_unused]] const clockid_t& clockid)
+    {
+#ifndef USE_HW_BREAKPOINT_COMPAT
+        attr_.use_clockid = use_clockid;
+        attr_.clockid = clockid;
+#endif
+    }
+
+    // When we poll on the fd given by perf_event_open, wakeup, when our buffer is 80% full
+    // Default behaviour is to wakeup on every event, which is horrible performance wise
+    void set_watermark(const size_t& mmap_pages)
+    {
+        attr_.watermark = 1;
+        attr_.wakeup_watermark = static_cast<uint32_t>(0.8 * mmap_pages * sysconf(_SC_PAGESIZE));
+    }
+
+    void set_exclude_kernel(const bool& exclude_kernel)
+    {
+        attr_.exclude_kernel = exclude_kernel;
+    }
+
+    void set_sample_period(const int& period);
+    void set_sample_freq(const uint64_t& freq);
+    void event_attr_update(std::uint64_t value, const std::string& format);
+
+    void parse_pmu_path(const std::string& ev_name);
+    void parse_cpus();
+    const std::set<Cpu>& supported_cpus() const;
+
+    bool is_valid() const;
+    bool event_is_openable();
+
+    bool is_available_in(ExecutionScope scope) const
+    {
+        // per-process should always work. the counter will just not count if the process is
+        // scheduled on a core that is not supprted by that counter
+        return scope.is_thread() || cpus_.empty() || cpus_.count(scope.as_cpu());
+    }
+
+    bool degrade_precision();
+
+    friend bool operator==(const Event& lhs, const Event& rhs)
+    {
+        return !memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator<(const Event& lhs, const Event& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator>(const Event& lhs, const Event& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
```

Comment:
these are identical, so two Event instances a and b are equal: a == b, or both
a < b and a > b.